ql/src/java/org/apache/hadoop/hive/ql/exec/HiveStructComparator.java

```diff
@@ -45,16 +48,14 @@ public int compare(Object key1, Object key2) {
         if (a1.size() == 0) {
             return 0;
         }
-        if (comparator == null) {
```

Comment:
The root cause is that the num of comparators is fixed to the size of the first List. This will work for STRUCT since the number of elements is consistent across records. However, in the case of ARRAY, it varies.